nixos/nix-daemon: use structural settings

polykernel opened the pull request:
The `nix.*` options, apart from options for setting up the
daemon itself, currently provide a lot of setting mappings
for the Nix daemon configuration. The scope of the mapping yields
convience, but the line where an option is considered essential
is blurry. For instance, the extra-sandbox-paths mapping is
provided without its primary consumer, and the corresponding
sandbox-paths option is also not mapped.

This increases the maintenance burden as maintainers have to
closely follow upstream changes which in the case of Nix is made
worse by having to juggle between two state versions, the stable release
and the unstable release, which may include or deprecate features.

One conditional option `publicHostKey` was added to the `buildMachines`
submodule corresponding to the newly added base64 encoded public host
key setting exposed in the builder syntax. The build machine generation
was subsequently rewritten to use concatStringsSep to group concatenations.

This commit aims to following the standard outlined in RFC 42[1] to
implement a structural setting pattern. The Nix configuration is encoded
at its core as key-value pairs which maps nicely to attribute sets, making
it feasible to express in the Nix language itself. Some existing options are
kept such as `buildMachines` and `registry` which present a simplified interface
to managing the respective settings.

Legacy configurations are mapped to their corresponding options under the
structural settings for backwards compatibility.

Various options settings in other nixos modules and relevant tests have been
updated to use the structural setting for consistency.

The generation and validation of the configration file has been modified to
use `writeTextFile` instead of `runCommand` for clarity. Note that validation
is now mandatory as strict checking of options has been pushed down to the
derivation level due to freeformType consuming unmatched options. Furthermore,
validation can not occur when cross-compiling due to current limitations.

[1] - https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
Initial effort for https://github.com/nix-community/home-manager/issues/2324

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

ShamrockLee commented on nixos/modules/services/misc/nix-daemon.nix:
```diff
@@ -498,38 +385,237 @@ in
+
+      settings = mkOption {
+        type = types.submodule {
+          freeformType = semanticConfType;
```
> but I am not sure if having the error/warning at the derivation level as opposed to the module system level is acceptable.

I'm still a novice here. IMHO, the modular configurations eventually comes down to the derivation that generates the `nix.conf` file. As the generation process itself is already outside (or underneath) the modular expressions, the vlidation of the output file `nix.conf` would not create an extra  "non-modulalized" part.

By the way, there are some modules who use the `generator` attribute provided by `type` to generate the output file, where the `type` attribute is specified under the `options` tree. Maybe the validation could be specified as part of the `type` or `options`, we want to go even more "modular". This would probably go through some "consesus-reaching" process such as an additional issue or an RFC, and modules using a rather imperative way of validation would then be modified.

polykernel replied:
> By the way, there are some modules who use the generator attribute provided by type to generate the output file, where the type attribute is specified under the options tree. Maybe the validation could be specified as part of the type or options to go even more "modular". 

Could you elaborate on the validation mechanism, I am not sure I understand correctly but do you mean using `lib.generators` to generate the config which itself and specifying a schema under an option which gets match to its value? The issue when I thought about it is using a schema which is analogous to the mapping all config options in the `options` tree is it has to be total and essentially match upstream, which increases maintenance burden and is what RFC 42 tries to avoid.

ShamrockLee replied:
I cannot recall where I saw them. Maybe I have mixed something up.

It seems that the generators produce strings instead of files. It then makes more sense to attach the validation through package overriding or similar approaches.

roberth replied:
@polykernel 

> not sure if having the error/warning at the derivation level as opposed to the module system level is acceptable.

While earlier is better, it is still early enough, as derivation builds are also free of side effects.

@ShamrockLee You're thinking of `pkgs.formats`. Not being in `lib`, those can produce a derivation. It's also a better name :)

> Maybe the validation could be specified as part of the type or options to go even more "modular". This would probably go through some "consesus-reaching" process such as an additional issue or an RFC, and modules using a rather imperative way of validation would then be modified to fit into the standardized pattern.

The separation we have right now, with formats producing a `type` that returns plain Nix values instead of a derivation, is quite useful because it allows modules to use values from the config. This wouldn't be possible if the type converted the settings to a derivation directly.
I suppose we could use the formats as 'circular programs' though, taking more advantage of laziness. See #144216

polykernel commented on nixos/modules/system/boot/binfmt.nix:
```diff
@@ -262,11 +262,11 @@ in {
       } // (magics.${system} or (throw "Cannot create binfmt registration for system ${system}"));
     }) cfg.emulatedSystems);
     # TODO: add a nix.extraPlatforms option to NixOS!
```
Updated, thanks.

polykernel committed:
nixos/nix-daemon: use structural settings

The `nix.*` options, apart from options for setting up the
daemon itself, currently provide a lot of setting mappings
for the Nix daemon configuration. The scope of the mapping yields
convience, but the line where an option is considered essential
is blurry. For instance, the `extra-sandbox-paths` mapping is
provided without its primary consumer, and the corresponding
`sandbox-paths` option is also not mapped.

The current system increases the maintenance burden as maintainers have to
closely follow upstream changes. In this case, there are two state versions
of Nix which have to be maintained collectively, with different options
avaliable.

This commit aims to following the standard outlined in RFC 42[1] to
implement a structural setting pattern. The Nix configuration is encoded
at its core as key-value pairs which maps nicely to attribute sets, making
it feasible to express in the Nix language itself. Some existing options are
kept such as `buildMachines` and `registry` which present a simplified interface
to managing the respective settings. The interface is exposed as `nix.settings`.

Legacy configurations are mapped to their corresponding options under `nix.settings`
for backwards compatibility.

Various options settings in other nixos modules and relevant tests have been
updated to use structural setting for consistency.

The generation and validation of the configration file has been modified to
use `writeTextFile` instead of `runCommand` for clarity. Note that validation
is now mandatory as strict checking of options has been pushed down to the
derivation level due to freeformType consuming unmatched options. Furthermore,
validation can not occur when cross-compiling due to current limitations.

A new option `publicHostKey` was added to the `buildMachines`
submodule corresponding to the base64 encoded public host key settings
exposed in the builder syntax. The build machine generation was subsequently
rewritten to use `concatStringsSep` for better performance by grouping
concatenations.

[1] - https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md

dasJ reviewed (approved):
Big 👍 for making the settings consistent with the nix manual and for using a much nicer way of specifying them. However I am not sure if this should go into the 21.11 release or if we should wait for 22.05. In any case we should probably wait for the branch-off and backport to 21.11 if necessary

ShamrockLee commented:
It has been a while after the 21.11 branch-off.

Looking forward to its merging!

thiagokokada commented:
Is there something blocking this merge?

(Except from the conflicts).

roberth commented:
I don't see any blockers and the conflict seems simple.

polykernel commented:
I have resolved the conflict and updated the `publicHostKey` option to be static.

ShamrockLee commented:
As the target changelog release note is 21.11, this is to be back-ported to `release-21.11`. Is that right?

It seems alright as the backward-compatibility is kept with the mapping layer.

dasJ commented:
Thanks for pointing this out, I was close to merging this ;)
I don't think we should backport this since it introduces a lot of warnings for existing setups and the large size of this PR leaves room for regressions.
If the release notes are moved to 22.05, I will happily merge this (unless @roberth or anyone else intervenes)

roberth commented on nixos/modules/services/continuous-integration/hydra/default.nix:
```diff
@@ -277,16 +275,21 @@ in
+      {
+        keep-outputs = true;
+        keep-derivations = true;
+        nix.settings.trusted-users = [ "hydra-queue-runner" ];
```
```suggestion
        trusted-users = [ "hydra-queue-runner" ];
```

Although hydra is about to be removed iiuc.

polykernel replied:
thanks!

ShamrockLee replied:
> hydra is about to be removed iiuc

How? (just an off-topic question)

roberth replied:
There's talk about removing the nixUnstable it depends on in favor of using flakes to install hydra. Maybe someone will maintain a "stable" hydra instead, if there's even such a thing.

roberth commented on nixos/tests/os-prober.nix:
```diff
@@ -43,11 +43,11 @@ let
-      hashed-mirrors =
-      connect-timeout = 1
-    '';
+    nix.settings = {
+      substituters = lib.mkForce [];
+      hashed-mirrors = null;
+      connect-timeout = 1;
+    };
```
There's a bunch of these. Should probably be the default for VM tests. Not a requirement for this PR though.

roberth commented on nixos/modules/services/misc/nix-daemon.nix:
```diff
@@ -579,44 +671,49 @@ in
+      let badMachine = m: m.system == null && m.systems == [ ];
+      in
+      [
         {
-          assertion = !(builtins.any badMachine cfg.buildMachines);
+          assertion = !(any badMachine cfg.buildMachines);
           message = ''
             At least one system type (via <varname>system</varname> or
               <varname>systems</varname>) must be set for every build machine.
               Invalid machine specifications:
           '' + "      " +
-          (builtins.concatStringsSep "\n      "
-            (builtins.map (m: m.hostName)
-              (builtins.filter (badMachine) cfg.buildMachines)));
+          (concatStringsSep "\n      "
+            (map (m: m.hostName)
+              (filter (badMachine) cfg.buildMachines)));
         }
       ];
```
This would benefit from a second attempt at submodule assertions, https://github.com/NixOS/nixpkgs/pull/97023#issuecomment-999639778, both in code complexity and error message quality.
(off-topic)

roberth commented on nixos/doc/manual/from_md/release-notes/rl-2111.section.xml:
```diff
@@ -2047,6 +2047,34 @@ Superuser created successfully.
           <literal>java-packages.compiler</literal>.
         </para>
       </listitem>
+      <listitem>
```
This needs to move to a prominent location in rl-2205

polykernel replied:
I have moved the relevant release notes to rl-2205.

roberth reviewed (changes requested):
Small things and off-topic things. Looking pretty good otherwise.

polykernel commented on nixos/modules/services/misc/nix-daemon.nix:
```diff
@@ -537,35 +430,234 @@ in
+          </citerefentry> for avalaible options.
+          The value declared here will be translated directly to the key-value pairs
+          Nix expects. This is much more flexible than the plethora of options
+          under <option>nix.*</option>, but lacks any type checking.
```
thanks, I tend to over-explain things a lot and I agree in this case the details are not necessary.

mweinelt commented:
After this change `nix.buildMachines`, which I did not change at all, is apparently failing.

```
error: value is a string while a list was expected

       at /etc/nixos/nixpkgs/nixos/modules/services/misc/nix-daemon.nix:676:14:

          675|           (machine:
          676|             (concatStringsSep " " [
             |              ^
          677|               "${optionalString (machine.sshUser != null) "${machine.sshUser}@"}${machine.hostName}"

       … while evaluating anonymous lambda

       at /etc/nixos/nixpkgs/nixos/modules/services/misc/nix-daemon.nix:675:12:

          674|         concatMapStrings
          675|           (machine:
             |            ^
          676|             (concatStringsSep " " [

       … from call site

       … while evaluating 'concatMapStrings'
```

lovesegfault commented:
Yup, this broke the evaluation for any/everyone using `nix.buildMachines`

polykernel commented:
I missed a bracket around the append operator. I apologize for thoroughly testing the changes enough.

roberth commented:
We should consider not warning about the renames until 21.11 is end of life. #157667 describes why sudden renames are a problem for out-of-tree modules.

wahjava commented:
The `s/warning/error/` and validation introduced in this commit is causing errors now:

```
building '/nix/store/d2wav40xyk3d13fpbzc3dn1d5249zp1s-etc-registry.json.drv'...
building '/nix/store/41y6qwy8v3n08navcmkgk34m9rjwqpmb-nix.conf.drv'...
error: builder for '/nix/store/41y6qwy8v3n08navcmkgk34m9rjwqpmb-nix.conf.drv' failed with exit code 1;
       last 2 log lines:
       > Validating generated nix.conf
       > error: unknown experimental feature 'ca-references'
       For full logs, run 'nix log /nix/store/41y6qwy8v3n08navcmkgk34m9rjwqpmb-nix.conf.drv'.
error: 1 dependencies of derivation '/nix/store/kydaidpymdirx9yxyfhrnwbfdk2pykj6-etc.drv' failed to build
error: 1 dependencies of derivation '/nix/store/7snhkayxbb1rpmqbhs8fgjwsmkdsmlpy-nixos-system-host-22.05.20220130.efeefb2.drv' failed to build
```

The only way is to disable `nix.checkConfig`.

polykernel commented:
The validation must catch all warnings and errors because the config checker does not distinguish between invalid syntax and invalid data. Admittedly this is not as flexible as the type checking in the module system but a necessity given the limitations of freeform modules.

In this case, it seems the `ca-references` experimental feature have been removed as per https://github.com/NixOS/nix/issues/5913#issuecomment-1013513176.

wahjava commented:
I don't have `ca-references` explicitly declared in my nix configuration, but still it's making it there somehow.

```
❯ fgrep ca- /etc/nix/nix.conf
extra-experimental-features = nix-command ca-references flakes
```

And my nix nixos configuration looks like:

```nix
{ inputs, pkgs, ...} : {
  nix = {

    gc = {
      automatic = true;
      dates = "13:37";
      options = "--delete-older-than 10d";
    };

    settings = {
      allowed-users = [ "@wheel" "root" ];
      sandbox = true;
      cores = 4;
      max-jobs = 1;
      substituters =  [
        "https://nix-community.cachix.org"
        "https://hydra.iohk.io"
        "https://nixpkgs-wayland.cachix.org"
      ];
      trusted-public-keys =  [
        "nix-community.cachix.org-1:mB9FSh9qf2dCimDSUo8Zy7bkq5CX+/rkCWyvRCYg3Fs="
        "hydra.iohk.io:f/Ea+s+dFdN+3Y/G+FDgSq+a5NEWhJGzdjvKNGv0/EQ="
        "nixpkgs-wayland.cachix.org-1:3lwxaILxMRkVhehr5StQprHdEo4IrE8sRho9R9HOLYA="
      ];
    };

    checkConfig = false; # workaround for https://github.com/NixOS/nixpkgs/pull/139075#issuecomment-1027772745

    nixPath = [ "nixpkgs=${inputs.nixpkgs}" ];

    package = pkgs.nix_2_6;

    extraOptions = ''
      experimental-features = nix-command flakes
      keep-outputs = true
      keep-derivations = true
    '';
  };
}
```

polykernel commented:
I cannot seem to reproduce this with the example configuration given in a vm.
```sh
> fgrep experimental-features /etc/nix/nix.conf
experimental-features = nix-command flakes
```

Could something else be writing to `nix.extraOptions` or `nix.settings`? You could test it by writing some debugging assertion or use `mkRemovedOptionModule` to get a trace of where it is called, but I am afraid such assertions wouldn't be able to distinguish merging of `nix.extraOptions`.

wahjava commented:
[Found it](https://github.com/gytis-ivaskevicius/flake-utils-plus/blob/master/lib/mkFlake.nix#L175), sorry.

roberth commented:
The `options` module argument contains info about which modules contribute to config values.

`nix repl '<nixpkgs>'`
```
nix-repl> s = (pkgs.nixos { }).options.nix.settings

nix-repl> :p lib.zipListsWith lib.nameValuePair s.files s.definitions        
[ { name = "/home/user/h/nixpkgs/nixos/modules/services/misc/nix-daemon.nix"; value = { substituters = [ "https://cache.nixos.org/" ]; system-features = { _type = "override"; content = [ "nixos-test" "benchmark" "big-parallel" "kvm" ]; priority = 1000; }; trusted-public-keys = [ "cache.nixos.org-1:6NCHdD59X431o0gWypbMrAURkbJ16ZPMQFGspcDShjY=" ]; }; } { name = "/home/user/h/nixpkgs/nixos/modules/services/misc/nix-daemon.nix"; value = { builders = null; }; } { name = "/home/user/h/nixpkgs/nixos/modules/services/misc/nix-daemon.nix"; value = { sandbox-fallback = false; }; } ]

```

You'll probably want to use your own config instead of `pkgs.nixos { }` which is an empty/default configuration except for `nixpkgs.pkgs`.

jian-lin commented on nixos/modules/services/misc/nix-daemon.nix:
```diff
@@ -579,44 +669,49 @@ in
+              (if machine.sshKey != null then machine.sshKey else "-")
+              (toString machine.maxJobs)
+              (toString machine.speedFactor)
+              (concatStringsSep "," machine.supportedFeatures)
```
see #159014

jian-lin commented:
Could anyone have a look at #159014 please?
